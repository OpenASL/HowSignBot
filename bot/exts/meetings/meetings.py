import asyncio
import logging
from enum import auto
from enum import Enum
from typing import cast
from typing import List
from typing import Optional
from typing import Union

import disnake
from disnake import ApplicationCommandInteraction
from disnake import MessageInteraction
from disnake.ext.commands import Bot
from disnake.ext.commands import check
from disnake.ext.commands import Cog
from disnake.ext.commands import command
from disnake.ext.commands import Context
from disnake.ext.commands import errors
from disnake.ext.commands import group
from disnake.ext.commands import is_owner
from disnake.ext.commands import Param
from disnake.ext.commands import slash_command

import meetings
from ._zoom import add_repost_after_delay
from ._zoom import get_zoom_meeting_id
from ._zoom import is_allowed_zoom_access
from ._zoom import make_zoom_embed
from ._zoom import REPOST_EMOJI
from ._zoom import ZOOM_CLOSED_MESSAGE
from ._zoom import zoom_impl
from ._zoom import ZoomCreateError
from bot import settings
from bot.database import store
from bot.utils.deprecation import send_deprecation_notice
from bot.utils.reactions import add_stop_sign
from bot.utils.reactions import get_reaction_message
from bot.utils.reactions import handle_close_reaction
from bot.utils.reactions import maybe_clear_reaction
from bot.utils.reactions import should_handle_reaction
from bot.utils.reactions import STOP_SIGN
from bot.utils.ui import ButtonGroupOption
from bot.utils.ui import ButtonGroupView
from bot.utils.ui import DropdownView

logger = logging.getLogger(__name__)

COMMAND_PREFIX = settings.COMMAND_PREFIX

MEET_CLOSED_MESSAGE = "‚ú® _Jitsi Meet ended_"
SPEAKEASY_CLOSED_MESSAGE = "‚ú® _Speakeasy event ended_"
WATCH2GETHER_HELP = """Start a watch2gether session

You can optionally pass a URL to use for the first video.

Examples:
```
{COMMAND_PREFIX}w2g
{COMMAND_PREFIX}w2g https://www.youtube.com/watch?v=DaMjr4AfYA0
```
""".format(
    COMMAND_PREFIX=COMMAND_PREFIX
)

WATCH2GETHER_CLOSED_MESSAGE = "‚ú® _watch2gether room closed_"


def make_jitsi_embed(meeting: meetings.JitsiMeet):
    title = f"<{meeting.join_url}>"
    content = (
        f"**Join URL**: <{meeting.join_url}>\n**Desktop App Link***: <{meeting.deeplink}>"
    )
    if meeting.name:
        content = f"{content}\n**Name**: {meeting.name}"
    content = f"{content}\n\nüöÄ This meeting is happening now. Go practice!\n*Desktop App Link requires <https://github.com/jitsi/jitsi-meet-electron>\n*After the meeting ends, click {STOP_SIGN} to remove this message.*"
    logger.info("sending jitsi meet info")
    return disnake.Embed(
        title=title,
        description=content,
        color=disnake.Color.blue(),
    )


def make_watch2gether_embed(url: str, video_url: Optional[str]) -> disnake.Embed:
    description = "üöÄ Watch videos together!"
    if video_url:
        description += f"\nQueued video: <{video_url}>"
    description += "\n*When finished, click üõë to remove this message.*"
    return disnake.Embed(title=url, description=description, color=disnake.Color.gold())


class ProtectionType(Enum):
    WAITING_ROOM = auto()
    FS_CAPTCHA = auto()


class Meetings(Cog):
    def __init__(self, bot: Bot):
        self.bot = bot

    @slash_command(name="zoom")
    @check(is_allowed_zoom_access)
    async def zoom_command(self, inter: ApplicationCommandInteraction):
        pass

    @zoom_command.sub_command(name="create")
    async def zoom_create(self, inter: ApplicationCommandInteraction):
        """(Authorized users only) Create a Zoom meeting"""
        assert inter.user is not None

        view = ButtonGroupView.from_options(
            options=(
                ButtonGroupOption(
                    label="FS Captcha", value=ProtectionType.FS_CAPTCHA, emoji="üëå"
                ),
                ButtonGroupOption(
                    label="Waiting Room", value=ProtectionType.WAITING_ROOM, emoji="üö™"
                ),
            ),
            creator_id=inter.user.id,
        )
        await inter.send(
            "üîê **How do you want to protect the meeting?** Choose one.", view=view
        )
        value = await view.wait_for_value()
        with_zzzzoom = value == ProtectionType.FS_CAPTCHA

        async def send_channel_message(mid: int):
            return await inter.channel.send(embed=await make_zoom_embed(mid))

        await zoom_impl(
            bot=self.bot,
            zoom_user=settings.ZOOM_USERS[inter.user.id],
            channel_id=inter.channel.id,
            meeting_id=None,
            send_channel_message=send_channel_message,
            set_up=True,
            with_zzzzoom=with_zzzzoom,
        )

    @zoom_command.sub_command(name="crosspost")
    async def zoom_crosspost(self, inter: ApplicationCommandInteraction, meeting_id: str):
        """Crosspost a previously-created Zoom meeting

        Parameters
        ----------
        meeting_id: Zoom meeting ID or zzzzoom ID
        """
        assert inter.user is not None

        zoom_or_zzzzoom_id: Union[int, str]
        try:
            zoom_or_zzzzoom_id = int(meeting_id)
        except ValueError:
            zoom_or_zzzzoom_id = meeting_id
            with_zzzzoom = True
        else:
            with_zzzzoom = False

        async def send_channel_message(mid: int):
            return await inter.channel.send(embed=await make_zoom_embed(mid))

        await zoom_impl(
            bot=self.bot,
            zoom_user=settings.ZOOM_USERS[inter.user.id],
            channel_id=inter.channel.id,
            meeting_id=zoom_or_zzzzoom_id,
            send_channel_message=send_channel_message,
            set_up=True,
            with_zzzzoom=with_zzzzoom,
        )
        await inter.send("_Crossposted Zoom_")

    @zoom_command.sub_command(name="stop")
    async def zoom_stop(
        self,
        inter: ApplicationCommandInteraction,
        meeting_id_str: str = Param(name="meeting_id"),
    ):
        """Remove meeting details for a Zoom meeting

        Parameters
        ----------
        meeting_id: Zoom meeting ID or zzzzoom ID
        """
        zoom_or_zzzzoom_id: Union[int, str]
        try:
            zoom_or_zzzzoom_id = int(meeting_id_str)
        except ValueError:
            zoom_or_zzzzoom_id = meeting_id_str

        meeting_id = await get_zoom_meeting_id(zoom_or_zzzzoom_id)
        meeting_exists = await store.zoom_meeting_exists(meeting_id=meeting_id)
        if not meeting_exists:
            raise errors.CheckFailure(
                f"‚ö†Ô∏è Could not find Zoom meeting with ID {meeting_id}. Make sure to run `{COMMAND_PREFIX}zoom setup {meeting_id}` first."
            )
        zoom_messages = tuple(await store.get_zoom_messages(meeting_id=meeting_id))
        if not zoom_messages:
            raise errors.CheckFailure(f"‚ö†Ô∏è No meeting messages for meeting {meeting_id}.")
        messages = []
        for message_info in zoom_messages:
            channel_id = message_info["channel_id"]
            message_id = message_info["message_id"]
            channel = self.bot.get_channel(channel_id)
            message: disnake.Message = await channel.fetch_message(message_id)
            messages.append(message)
            logger.info(
                f"scrubbing meeting details for meeting {meeting_id} in channel {channel_id}, message {message_id}"
            )
            await message.edit(content=ZOOM_CLOSED_MESSAGE, embed=None)
            await maybe_clear_reaction(message, REPOST_EMOJI)
        await store.end_zoom_meeting(meeting_id=meeting_id)
        await inter.send("üõë Meeting details removed.")

    # XXX: Ideally we'd use slash command permissions intead of a check here,
    # but those can only be set per-guild at the moment.
    @zoom_command.sub_command(name="users")
    @is_owner()
    async def zoom_users(self, inter: ApplicationCommandInteraction):
        """(Bot owner only) List users who have access to the zoom commands"""
        try:
            users = await meetings.list_zoom_users(token=settings.ZOOM_JWT)
        except asyncio.exceptions.TimeoutError:
            logger.exception("zoom request timed out")
            await inter.send(
                "üö® _Request to Zoom API timed out. This may be due to rate limiting. Try again later._"
            )
            return
        licensed_user_emails = {
            user.email for user in users if user.type == meetings.ZoomPlanType.LICENSED
        }
        description = "\n".join(
            tuple(
                ("üëë " if email.lower() in licensed_user_emails else "") + f"<@!{user_id}>"
                for user_id, email in settings.ZOOM_USERS.items()
            )
        )
        embed = disnake.Embed(
            title=f"Fancy Zoom Users ({len(settings.ZOOM_USERS)})",
            description=description,
            color=disnake.Color.blue(),
        )
        embed.set_footer(text="üëë = Licensed")
        await inter.send(embed=embed)

    @zoom_command.sub_command(
        name="license", hidden=True, help="Upgrade a user to the Licensed plan type."
    )
    @is_owner()
    async def zoom_license(
        self,
        inter: ApplicationCommandInteraction,
        user: disnake.User,
    ):
        """(Bot owner only) Upgrade a Zoom user to a Licensed plan"""
        assert inter.user is not None
        if user.id not in settings.ZOOM_USERS:
            await inter.send(f"üö® _{user.mention} is not a configured Zoom user._")
            return

        await inter.send(
            f"‚úã **{user.mention} will be upgraded to a Licensed plan**.",
        )
        zoom_user_id = settings.ZOOM_USERS[user.id]
        try:
            logger.info(f"attempting to upgrade user {user.id} to licensed plan")
            await meetings.update_zoom_user(
                token=settings.ZOOM_JWT,
                user_id=zoom_user_id,
                data={"type": meetings.ZoomPlanType.LICENSED},
            )
        except meetings.MaxZoomLicensesError:
            try:
                users = await meetings.list_zoom_users(token=settings.ZOOM_JWT)
            except asyncio.exceptions.TimeoutError:
                logger.exception("zoom request timed out")
                await inter.send(
                    "üö® _Request to Zoom API timed out. This may be due to rate limiting. Try again later._"
                )
                return
            zoom_to_discord_user_mapping = {
                email.lower(): disnake_id
                for disnake_id, email in settings.ZOOM_USERS.items()
            }
            # Discord user IDs for Licensed users
            licensed_user_discord_ids = tuple(
                zoom_to_discord_user_mapping[user.email.lower()]
                for user in users
                if user.email.lower() in zoom_to_discord_user_mapping
                and user.type == meetings.ZoomPlanType.LICENSED
                # Don't allow de-licensing the bot owner, of course
                and zoom_to_discord_user_mapping[user.email.lower()] != settings.OWNER_ID
            )
            if len(licensed_user_discord_ids):
                options = [
                    disnake.SelectOption(
                        label=settings.ZOOM_USERS[discord_user_id], value=discord_user_id
                    )
                    for discord_user_id in licensed_user_discord_ids
                ]

                async def on_select(select_interaction: MessageInteraction, value: str):
                    downgraded_user_id = int(value)
                    await select_interaction.response.edit_message(
                        content=f"‚òëÔ∏è Selected <@!{downgraded_user_id}> to downgrade.",
                        view=None,
                    )
                    try:
                        logger.info(
                            f"attempting to downgrade user {downgraded_user_id} to basic plan"
                        )
                        await meetings.update_zoom_user(
                            token=settings.ZOOM_JWT,
                            user_id=settings.ZOOM_USERS[downgraded_user_id],
                            data={"type": meetings.ZoomPlanType.BASIC},
                        )
                    except meetings.ZoomClientError:
                        logger.exception(f"failed to downgrade user {downgraded_user_id}")
                        await inter.send(
                            f"üö® _Failed to downgrade <@!{downgraded_user_id}>. Check the logs for details._"
                        )
                    try:
                        logger.info(
                            f"re-attempting to upgrade user {user.id} to licensed plan"
                        )
                        await meetings.update_zoom_user(
                            token=settings.ZOOM_JWT,
                            user_id=zoom_user_id,
                            data={"type": meetings.ZoomPlanType.LICENSED},
                        )
                    except meetings.ZoomClientError:
                        logger.exception(f"failed to upgrade user {user.id}")
                        await inter.send(
                            f"üö® _Failed to upgrade {user.mention}. Check the logs for details._"
                        )
                    await inter.send(
                        f"üëë **{user.mention} successfully upgraded to Licensed plan.**\n<@!{downgraded_user_id}> downgraded to Basic."
                    )

                view = DropdownView.from_options(
                    options=options, on_select=on_select, placeholder="Choose a user"
                )
                await inter.send("Choose a user to downgrade to Basic.", view=view)
            else:
                await inter.send(
                    "üö® _No available users to downgrade on Discord. Go to the Zoom account settings to manage licenses_."
                )
                return
            return
        except meetings.ZoomClientError as error:
            await inter.send(f"üö® _{error.args[0]}_")
            return
        except Exception:
            logger.exception(f"failed to license user {user}")
            await inter.send(
                f"üö® _Failed to license user {user.mention}. Check the logs for details._"
            )
            return
        else:
            await inter.send(f"üëë **{user.mention} upgraded to a Licensed plan**.")

    # Deprecated prefix commands

    @group(name="zoom", aliases=("z",), invoke_without_command=True)
    @check(is_allowed_zoom_access)
    async def zoom_group(
        self, ctx: Context, meeting_id: Optional[Union[int, str]] = None
    ):
        """AUTHORIZED USERS ONLY: Start a Zoom meeting"""
        await self.zoom_group_impl(ctx, meeting_id=meeting_id, with_zzzzoom=False)

    @zoom_group.command(
        name="setup",
        help="Set up a Zoom before revealing its details to other users. Useful for meetings that have breakout rooms.",
    )
    @check(is_allowed_zoom_access)
    async def zoom_setup(
        self, ctx: Context, meeting_id: Optional[Union[int, str]] = None
    ):
        await self.zoom_setup_impl(ctx, meeting_id=meeting_id, with_zzzzoom=False)

    @zoom_group.command(
        name="start",
        help="Reveal meeting details for a meeting started with the setup command",
    )
    @check(is_allowed_zoom_access)
    async def zoom_setup_start(
        self, ctx: Context, meeting_id: Optional[Union[int, str]] = None
    ):
        await ctx.channel.trigger_typing()
        if meeting_id:
            meeting_id = await get_zoom_meeting_id(meeting_id)
            meeting_exists = await store.zoom_meeting_exists(meeting_id=meeting_id)
            if not meeting_exists:
                raise errors.CheckFailure(
                    f"‚ö†Ô∏è Could not find Zoom meeting with ID {meeting_id}. Make sure to run `{COMMAND_PREFIX}zoom setup {meeting_id}` first."
                )
        else:
            zoom_user = settings.ZOOM_USERS[ctx.author.id]
            latest_meeting = await store.get_latest_pending_zoom_meeting_for_user(
                zoom_user
            )
            if not latest_meeting:
                raise errors.CheckFailure(
                    f"‚ö†Ô∏è You do not have any pending Zoom meetings. Make sure to run `{COMMAND_PREFIX}zoom setup [meeting_id]` first."
                )
            meeting_id = cast(int, latest_meeting["meeting_id"])
        await store.set_up_zoom_meeting(meeting_id=meeting_id)
        zoom_messages = tuple(await store.get_zoom_messages(meeting_id=meeting_id))
        if not zoom_messages:
            raise errors.CheckFailure(f"‚ö†Ô∏è No meeting messages for meeting {meeting_id}.")
        embed = await make_zoom_embed(meeting_id=meeting_id)
        messages: List[disnake.Message] = []
        for message_info in zoom_messages:
            channel_id = message_info["channel_id"]
            message_id = message_info["message_id"]
            channel = self.bot.get_channel(channel_id)
            message: disnake.Message = await channel.fetch_message(message_id)
            messages.append(message)
            logger.info(
                f"revealing meeting details for meeting {meeting_id} in channel {channel_id}, message {message_id}"
            )
            await message.edit(embed=embed)
            add_repost_after_delay(self.bot, message)
        if ctx.guild is None:
            links = "\n".join(
                f"[{message.guild} - #{message.channel}]({message.jump_url})"
                for message in messages
            )
            await ctx.send(
                embed=disnake.Embed(title="üöÄ Meeting Details Revealed", description=links)
            )
        else:
            channel_message = next(
                (message for message in messages if message.channel.id == ctx.channel.id),
                None,
            )
            if channel_message:
                await channel_message.reply(
                    f"üöÄ Meeting details revealed: {channel_message.jump_url}"
                )
            else:
                await ctx.channel.send("üöÄ Meeting details revealed.")

    @group(
        name="zzzzoom",
        aliases=("zzoom", "zzzoom", "zzzzzoom", "zz", "zzz", "zzzz", "zzzzz"),
        invoke_without_command=True,
    )
    @check(is_allowed_zoom_access)
    async def zzzzoom_group(
        self, ctx: Context, meeting_id: Optional[Union[str, int]] = None
    ):
        """AUTHORIZED USERS ONLY: Start a Zoom meeting and display the zzzzoom.us join URL instead of the normal join URL."""
        await self.zoom_group_impl(ctx, meeting_id=meeting_id, with_zzzzoom=True)

    @zzzzoom_group.command(
        name="setup",
        help="Set up a Zoom before revealing its details to other users. Useful for meetings that have breakout rooms.",
    )
    @check(is_allowed_zoom_access)
    async def zzzzoom_setup(
        self, ctx: Context, meeting_id: Optional[Union[int, str]] = None
    ):
        await self.zoom_setup_impl(ctx, meeting_id=meeting_id, with_zzzzoom=True)

    @zoom_command.error
    @zzzzoom_group.error
    @zzzzoom_setup.error
    @zoom_group.error
    @zoom_setup.error
    async def zoom_error(self, ctx, error):
        if isinstance(error, ZoomCreateError):
            logger.error("could not create zoom due to unexpected error", exc_info=error)
            await ctx.send(error)

    async def zoom_group_impl(
        self, ctx: Context, *, meeting_id: Optional[Union[int, str]], with_zzzzoom: bool
    ):
        await ctx.channel.trigger_typing()

        async def send_channel_message(mid: int):
            return await ctx.reply(embed=await make_zoom_embed(mid))

        await zoom_impl(
            bot=self.bot,
            zoom_user=settings.ZOOM_USERS[ctx.author.id],
            channel_id=ctx.channel.id,
            meeting_id=meeting_id,
            send_channel_message=send_channel_message,
            set_up=True,
            with_zzzzoom=with_zzzzoom,
        )

        before_example = f"{COMMAND_PREFIX}{ctx.invoked_with}"
        after_example = "/zoom"
        if meeting_id:
            substitute = "/zoom crosspost"
            before_example += " <meeting id>"
            after_example += " crosspost <meeting id>"
        else:
            substitute = "/zoom create"
            after_example += " create"
        await send_deprecation_notice(
            ctx,
            substitute=substitute,
            before_example=before_example,
            after_example=after_example,
        )

    async def zoom_setup_impl(
        self, ctx: Context, meeting_id: Optional[Union[int, str]], with_zzzzoom: bool
    ):
        await ctx.channel.trigger_typing()

        async def send_channel_message(_):
            return await ctx.reply(
                embed=disnake.Embed(
                    color=disnake.Color.blue(),
                    title="‚úã Stand By",
                    description="Zoom details will be posted here when the meeting is ready to start.",
                )
            )

        meeting_id, _ = await zoom_impl(
            bot=self.bot,
            zoom_user=settings.ZOOM_USERS[ctx.author.id],
            channel_id=ctx.channel.id,
            meeting_id=meeting_id,
            send_channel_message=send_channel_message,
            set_up=False,
            with_zzzzoom=with_zzzzoom,
        )

        zoom_messages = tuple(await store.get_zoom_messages(meeting_id=meeting_id))
        # DM zoom link and instructions once
        if len(zoom_messages) <= 1:
            command_name = "zzzzoom" if with_zzzzoom else "zoom"
            await ctx.author.send(
                content="üî® Set up your meeting below",
                embed=await make_zoom_embed(meeting_id, include_instructions=False),
            )
            await ctx.author.send(
                "To post in another channel, send the following command in that channel:\n"
                f"```{COMMAND_PREFIX}{command_name} setup {meeting_id}```\n"
                "When you're ready for people to join, reply with:\n"
                f"```{COMMAND_PREFIX}zoom start {meeting_id}```"
            )

    # End deprecated prefix commands

    @command(name="meet", aliases=("jitsi",), help="Start a Jitsi Meet meeting")
    async def meet_command(self, ctx: Context, *, name: Optional[str]):
        meeting = meetings.create_jitsi_meet(name, secret=settings.SECRET_KEY)
        logger.info("sending jitsi meet info")
        message = await ctx.send(embed=make_jitsi_embed(meeting))

        await add_stop_sign(message)

    @command(name="speakeasy", help="Start a Speakeasy (https://speakeasy.co/) event")
    async def speakeasy_command(self, ctx: Context, *, name: Optional[str]):
        join_url = meetings.create_speakeasy(name, secret=settings.SECRET_KEY)
        content = f"Ô∏èüçª **Speakeasy**\nJoin URL: <{join_url}>"
        if name:
            content = f"{content}\n**Name**: {name}"
        content = f"{content}\nüöÄ This event is happening now. Make a friend!"
        logger.info("sending speakeasy info")
        message = await ctx.send(content=content)
        await add_stop_sign(message)

    @command(name="w2g", aliases=("wtg", "watch2gether"), help=WATCH2GETHER_HELP)
    async def watch2gether_command(self, ctx: Context, video_url: Optional[str] = None):
        logger.info("creating watch2gether meeting")
        try:
            url = await meetings.create_watch2gether(
                settings.WATCH2GETHER_API_KEY, video_url
            )
        except Exception:
            logger.exception("could not create watch2gether room")
            message = await ctx.send(
                content="üö® _Could not create watch2gether room. That's embarrassing._"
            )
        else:
            message = await ctx.send(embed=make_watch2gether_embed(url, video_url))

        await add_stop_sign(message)

    async def edit_meeting_moved(self, message: disnake.Message) -> None:
        await message.edit(
            content=f"{REPOST_EMOJI} *Meeting details moved below.*",
            embed=None,
        )
        await maybe_clear_reaction(message, REPOST_EMOJI)

    @Cog.listener()
    async def on_raw_reaction_add(self, payload: disnake.RawReactionActionEvent) -> None:
        if should_handle_reaction(self.bot, payload, {REPOST_EMOJI}):
            message = await get_reaction_message(self.bot, payload)
            if not message:
                return
            zoom_message = await store.get_zoom_message(message.id)
            if not zoom_message:
                return
            zoom_meeting = await store.get_zoom_meeting(zoom_message["meeting_id"])
            if not zoom_meeting:
                return
            # Meeting isn't set up, don't reveal it yet
            if not zoom_meeting["setup_at"]:
                return
            original_message = None
            if message.reference:
                if message.reference.cached_message:
                    original_message = message.reference.cached_message
                else:
                    channel = self.bot.get_channel(message.channel.id)
                    original_message = await channel.fetch_message(
                        message.reference.message_id
                    )
            # Try to remove the old message and reply with a new message
            if original_message:
                try:
                    await message.delete()
                except Exception:
                    await self.edit_meeting_moved(message)
            else:
                await self.edit_meeting_moved(message)

            send_method = original_message.reply if original_message else message.reply
            new_message = await send_method(
                content="üëê **This meeting is still going**. Come on in!",
                embed=await make_zoom_embed(zoom_message["meeting_id"]),
                mention_author=False,
            )
            add_repost_after_delay(self.bot, new_message)

            async with store.transaction():
                await store.create_zoom_message(
                    message_id=new_message.id,
                    channel_id=new_message.channel.id,
                    meeting_id=zoom_message["meeting_id"],
                )
                await store.remove_zoom_message(message_id=zoom_message["message_id"])

            return

        async def close_zoom_message(msg: disnake.Message):
            await store.remove_zoom_message(message_id=msg.id)
            await maybe_clear_reaction(msg, REPOST_EMOJI)
            return ZOOM_CLOSED_MESSAGE

        await handle_close_reaction(
            self.bot,
            payload,
            close_messages={
                r"zoom\.us|Stand By|Could not create Zoom": close_zoom_message,
                r"meet\.jit\.si": MEET_CLOSED_MESSAGE,
                r"Speakeasy": SPEAKEASY_CLOSED_MESSAGE,
                r"w2g\.tv|Could not create watch2gether": WATCH2GETHER_CLOSED_MESSAGE,
            },
        )


def setup(bot: Bot) -> None:
    bot.add_cog(Meetings(bot))
