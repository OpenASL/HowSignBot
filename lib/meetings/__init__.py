from __future__ import annotations

import base64
import hashlib
import hmac
from collections.abc import AsyncIterator
from contextlib import asynccontextmanager
from enum import IntEnum
from typing import Callable, NamedTuple

import aiohttp
import cuteid
from slugify import slugify


class ZoomClientError(Exception):
    pass


class MaxZoomLicensesError(ZoomClientError):
    pass


class BasicUserLimitReachedError(ZoomClientError):
    pass


@asynccontextmanager
async def zoom_request(
    method: str,
    path: str,
    *,
    token: str,
    raise_for_status: bool = True,
    timeout: int = 10,
    **kwargs,
) -> AsyncIterator[aiohttp.ClientResponse]:
    timeout = aiohttp.ClientTimeout(total=timeout)
    async with aiohttp.request(
        method,
        f"https://api.zoom.us/v2/{path.lstrip('/')}",
        timeout=timeout,
        headers={"Authorization": f"Bearer {token}"},
        **kwargs,
    ) as resp:
        if raise_for_status:
            resp.raise_for_status()
        yield resp


class ZoomMeeting(NamedTuple):
    id: int
    join_url: str
    passcode: str
    topic: str


async def create_zoom(
    *, token: str, user_id: str, topic: str, settings: dict
) -> ZoomMeeting:
    """Create and return a Zoom meeting via the Zoom API."""
    async with zoom_request(
        "POST",
        f"/users/{user_id}/meetings",
        token=token,
        json={
            "type": 1,
            "topic": topic,
            "settings": settings,
        },
    ) as resp:
        data = await resp.json()
    return ZoomMeeting(
        id=data["id"],
        join_url=data["join_url"],
        passcode=data["password"],
        # Pass topic directly so we don't get the default 'Zoom Meeting' topic
        topic=topic,
    )


class ZoomPlanType(IntEnum):
    BASIC = 1
    LICENSED = 2


class ZoomUser(NamedTuple):
    id: str
    email: str
    type: ZoomPlanType


async def list_zoom_users(*, token: str) -> list[ZoomUser]:
    async with zoom_request("GET", "/users", token=token) as resp:
        data = await resp.json()
    return [ZoomUser(id=u["id"], email=u["email"], type=u["type"]) for u in data["users"]]


async def update_zoom_user(*, token: str, user_id: str, data: dict):
    async with zoom_request(
        "PATCH",
        f"/users/{user_id}",
        token=token,
        json=data,
        raise_for_status=False,
    ) as resp:
        if resp.status >= 400:
            resp_data = await resp.json()
            error_code = int(resp_data["code"])
            message = resp_data["message"]
            # NOTE: Contrary to Zoom's docs, 3412 is used for max licenses exceeded
            if error_code in {2034, 3412}:
                raise MaxZoomLicensesError(message)
            elif error_code == 2033:
                raise BasicUserLimitReachedError(message)
            else:
                resp.raise_for_status()
        else:
            resp.raise_for_status()
    return None


async def get_zoom(
    *,
    token: str,
    meeting_id: int,
) -> ZoomMeeting:
    """Get an existing Zoom meeting via the Zoom API."""
    async with aiohttp.ClientSession() as client:
        resp = await client.get(
            f"https://api.zoom.us/v2/meetings/{meeting_id}",
            headers={"Authorization": f"Bearer {token}"},
        )
    resp.raise_for_status()
    data = await resp.json()
    return ZoomMeeting(
        id=data["id"],
        join_url=data["join_url"],
        passcode=data["password"],
        topic=data["topic"],
    )


async def create_watch2gether(api_key: str, video_url: str | None = None) -> str:
    """Create and return a watch2gether URL via the watch2gether API."""
    async with aiohttp.ClientSession() as client:
        payload = {"w2g_api_key": api_key, "share": video_url}
        resp = await client.post("https://w2g.tv/rooms/create.json", json=payload)
    resp.raise_for_status()
    data = await resp.json()
    stream_key = data["streamkey"]
    url = f"https://w2g.tv/rooms/{stream_key}"
    return url


def _signature(s: str, *, secret: str) -> str:
    dig = hmac.new(
        secret.encode("utf-8"), msg=s.encode("utf-8"), digestmod=hashlib.sha256
    ).digest()
    return base64.urlsafe_b64encode(dig).decode()


def _slug_with_signature(s: str, *, secret: str, signature_length=16):
    return "-".join((slugify(s), _signature(s, secret=secret)[:signature_length]))


def _get_secret_slug(
    name: str | None, secret: str, fallback: Callable = cuteid.cuteid
) -> str:
    """Return a hard-to-guess slug to use for meeting URLs.

    If name is passed, return the slugified name with a signature
    appended so that the same meeting can be shared in multiple servers.
    """
    return _slug_with_signature(name, secret=secret) if name else fallback()


class JitsiMeet(NamedTuple):
    join_url: str
    deeplink: str
    name: str | None


def create_jitsi_meet(name: str | None, *, secret: str) -> JitsiMeet:
    """Return a Jitsi Meet URL."""
    slug = _get_secret_slug(name, secret)
    return JitsiMeet(
        join_url=f"https://meet.jit.si/{slug}", deeplink=f"jitsi-meet://{slug}", name=name
    )


def create_speakeasy(name: str | None, *, secret: str) -> str:
    """Return a Speakeasy URL."""
    slug = _get_secret_slug(name, secret, fallback=cuteid.emojid)
    return f"https://speakeasy.co/{slug}"
