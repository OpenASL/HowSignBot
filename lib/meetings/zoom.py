import base64
import time
from collections.abc import AsyncIterator
from contextlib import asynccontextmanager
from enum import IntEnum
from typing import NamedTuple

import aiohttp


class ZoomTokenManager:
    """Utility class for getting and storing server-to-server access tokens via
    via the OAuth token API.

    https://developers.zoom.us/docs/internal-apps/s2s-oauth/#use-account-credentials-to-get-an-access-token
    """

    URL = "https://zoom.us/oauth/token"

    def __init__(
        self,
        *,
        account_id: str,
        client_id: str,
        client_secret: str,
        renew_pad_secs: int = 60,
    ):
        self.account_id = account_id
        self.client_id = client_id
        self.client_secret = client_secret
        self.renew_pad_secs = renew_pad_secs
        self._token = None
        self._exp = None

    async def login(self):
        """Use account credentials to get and store an access token."""
        credentials = base64.b64encode(
            f"{self.client_id}:{self.client_secret}".encode()
        ).decode("utf-8")
        async with aiohttp.ClientSession() as client:
            resp = await client.post(
                self.URL,
                params={
                    "grant_type": "account_credentials",
                    "account_id": self.account_id,
                },
                headers={
                    "Host": "zoom.us",
                    "Authorization": f"Basic {credentials}",
                },
            )
        resp.raise_for_status()
        data = await resp.json()

        self._token = data["access_token"]
        self._exp = time.time() + data["expires_in"] - self.renew_pad_secs

    async def token(self):
        if not self._token or not self._exp or time.time() > self._exp:
            await self.login()

        return self._token


# -----------------------------------------------------------------------------


class ZoomMeeting(NamedTuple):
    id: int
    join_url: str
    passcode: str
    topic: str


class ZoomPlanType(IntEnum):
    BASIC = 1
    LICENSED = 2


class ZoomUser(NamedTuple):
    id: str
    email: str
    type: ZoomPlanType


# -----------------------------------------------------------------------------


class ZoomClient:
    def __init__(
        self,
        *,
        account_id: str,
        client_id: str,
        client_secret: str,
    ):
        self.token_manager = ZoomTokenManager(
            account_id=account_id, client_id=client_id, client_secret=client_secret
        )

    @asynccontextmanager
    async def _zoom_request(
        self,
        method: str,
        path: str,
        *,
        raise_for_status: bool = True,
        timeout: int = 10,
        **kwargs,
    ) -> AsyncIterator[aiohttp.ClientResponse]:
        client_timeout = aiohttp.ClientTimeout(total=timeout)
        token = await self.token_manager.token()
        async with aiohttp.request(
            method,
            f"https://api.zoom.us/v2/{path.lstrip('/')}",
            timeout=client_timeout,
            headers={"Host": "api.zoom.us", "Authorization": f"Bearer {token}"},
            **kwargs,
        ) as resp:
            if raise_for_status:
                resp.raise_for_status()
            yield resp

    async def create_zoom(
        self, *, user_id: str, topic: str, settings: dict
    ) -> ZoomMeeting:
        """Create and return a Zoom meeting via the Zoom API."""
        async with self._zoom_request(
            "POST",
            f"/users/{user_id}/meetings",
            json={
                "type": 1,
                "topic": topic,
                "settings": settings,
            },
        ) as resp:
            data = await resp.json()
        return ZoomMeeting(
            id=data["id"],
            join_url=data["join_url"],
            passcode=data["password"],
            # Pass topic directly so we don't get the default 'Zoom Meeting' topic
            topic=topic,
        )

    async def get_zoom(
        self,
        *,
        meeting_id: int,
    ) -> ZoomMeeting:
        """Get an existing Zoom meeting via the Zoom API."""
        async with self._zoom_request(
            "GET",
            f"/meetings/{meeting_id}",
        ) as resp:
            data = await resp.json()
        return ZoomMeeting(
            id=data["id"],
            join_url=data["join_url"],
            passcode=data["password"],
            topic=data["topic"],
        )

    async def list_zoom_users(self) -> list[ZoomUser]:
        async with self._zoom_request("GET", "/users") as resp:
            data = await resp.json()
        return [
            ZoomUser(id=u["id"], email=u["email"], type=u["type"]) for u in data["users"]
        ]
