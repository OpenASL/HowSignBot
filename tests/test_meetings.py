import meetings

SECRET = "myprecious"


def test_create_jitsi_meet_deterministic_urls(snapshot):
    name = "Practice...PRACTICE!!!PrAcTice"
    first = meetings.create_jitsi_meet(name, secret=SECRET)
    assert first == snapshot
    second = meetings.create_jitsi_meet(name, secret=SECRET)
    assert first == second
