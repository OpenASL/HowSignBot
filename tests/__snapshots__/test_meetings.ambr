# name: test_create_jitsi_meet_deterministic_urls
  <class 'JitsiMeet'> (
    deeplink='jitsi-meet://practice-practice-practice-X0rhbi4w1bXFwd5H',
    join_url='https://meet.jit.si/practice-practice-practice-X0rhbi4w1bXFwd5H',
  )
---
